app/src/org/commcare/views/ViewUtil.java

```diff
@@ -15,27 +15,30 @@
 import android.view.View;
 import android.view.inputmethod.InputMethodManager;
 
+import org.commcare.suite.model.Action;
 import org.commcare.suite.model.DisplayData;
 import org.commcare.utils.MediaUtil;
 import org.javarosa.core.services.locale.Localizer;
 
 import java.util.ArrayList;
 
 /**
- * Utilities for converting CommCare UI diplsay details into Android objects
+ * Utilities for converting CommCare UI display details into Android objects
  *
  * @author ctsims
  */
 public final class ViewUtil {
 
-    // This is silly and isn't really what we want here, but it's a start.
-    // (We'd like to be able to add a displayunit to a menu in a super
-    // easy/straightforward way.
-    public static void addDisplayToMenu(Context context, Menu menu,
-                                        int menuId, int menuGroupId, DisplayData display) {
+    public static void addActionToMenu(Context context, Action action, Menu menu, int menuId,
+                                       int menuGroupId) {
+        DisplayData display = action.getDisplay().evaluate();
         MenuItem item = menu.add(menuGroupId, menuId, menuId,
                 Localizer.clearArguments(display.getName()).trim());
-        if (display.getImageURI() != null) {
+        if (action.hasActionBarIcon() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
+            Bitmap b = MediaUtil.inflateDisplayImage(context, action.getActionBarIconReference());
+            item.setIcon(new BitmapDrawable(context.getResources(), b));
+            item.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS);
```

Comment:
I have slight reservations on the use of `SHOW_AS_ACTION_ALWAYS` here. The alternative is `SHOW_AS_ACTION_IF_ROOM`, the difference between the 2 being what it sounds like. I wanted to use the ALWAYS field here so that these actions would take precedence in the app bar over the search and barcode scan icons in the case of limited room (which I think they should, since the app config is saying it's important enough to its workflow to be obvious to the user). Note that any icons that get pushed out of the app bar still show up in the options menu, they are just not immediately visible. The downside of this is that if a user were to add several actions with an `action-bar-icon` in the same detail, they would end up looking very cluttered. Since the probability of that is low, and the resolution path would be pretty obvious to app builders (including fewer actions in the app bar), I think this is OK, but just wanted to flag it. This is an example of what it would look like for the custom actions to push the search icon off of the app bar and into the options menu:
![screenshot_20160921-100606](https://cloud.githubusercontent.com/assets/3723143/18715047/d01efa32-7fe5-11e6-93f5-2eb2e50ac431.png)